src/runtime_src/xdp/profile/device/traceS2MM.cpp

```diff
@@ -60,6 +107,124 @@ void TraceS2MM::showProperties()
+    result.HostTimestamp = result.HostTimestamp | partial;
+
+    if (mod == 3 && out_stream) {
+      (*out_stream) << std::hex
+      << "Clock Training sample : " << result.HostTimestamp << " " << result.Timestamp
+      << std::dec << std::endl;
+    }
+}
+
+inline void TraceS2MM::parsePacket(uint64_t packet, uint64_t firstTimestamp, xclTraceResults &result)
+{
+    result.Timestamp = (packet & 0x1FFFFFFFFFFF) - firstTimestamp;
+    result.EventType = ((packet >> 45) & 0xF) ? XCL_PERF_MON_END_EVENT :
+        XCL_PERF_MON_START_EVENT;
+    result.TraceID = (packet >> 49) & 0xFFF;
+    result.Reserved = (packet >> 61) & 0x1;
+    result.Overflow = (packet >> 62) & 0x1;
+    result.Error = (packet >> 63) & 0x1;
+    result.EventID = XCL_PERF_MON_HW_EVENT;
+    result.EventFlags = ((packet >> 45) & 0xF) | ((packet >> 57) & 0x10);
+
+    if (out_stream) {
+    static uint64_t previousTimestamp = 0;
+    (*out_stream) << std::dec << std::setw(5)
+        << "  Trace sample " << ": "
+        << dec2bin(uint32_t(packet>>32)) << " " << dec2bin(uint32_t(packet&0xFFFFFFFF)) << std::endl
+        << " Timestamp : " << result.Timestamp << "   "
+        << "Type : " << result.EventType << "   "
+        << "ID : " << result.TraceID << "   "
+        << "Pulse : " << static_cast<int>(result.Reserved) << "   "
+        << "Overflow : " << static_cast<int>(result.Overflow) << "   "
+        << "Err : " << static_cast<int>(result.Error) << "   "
+        << "Flags : " << static_cast<int>(result.EventFlags) << "   "
+        << "Interval : " << result.Timestamp - previousTimestamp << "   "
+        << std::endl;
+        previousTimestamp = result.Timestamp;
+    }
+}
+
+void TraceS2MM::parseTraceBuf(void* buf, uint64_t size, xclTraceResultsVector& traceVector)
+{
+    unsigned packetSizeBytes = 8;
+    unsigned tvindex = 0;
+    traceVector.mLength = 0;
+
+    auto count = size / packetSizeBytes;
+    if (count > MAX_TRACE_NUMBER_SAMPLES) {
+      count = MAX_TRACE_NUMBER_SAMPLES;
+    }
+    auto pos = static_cast<uint64_t*>(buf);
+    for (unsigned int i = 0; i < count; i++) {
+      auto currentPacket = pos[i];
+      if (!currentPacket)
+        return;
+      // Poor man's reset
+      if (i == 0 && !mPacketFirstTs)
+        mPacketFirstTs = currentPacket & 0x1FFFFFFFFFFF;
+      if (i < 8 && !mclockTrainingdone) {
+        unsigned int mod = i % 4;
+        parsePacketClockTrain(currentPacket, mPacketFirstTs, mod, traceVector.mArray[tvindex]);
+        tvindex = (mod == 3) ? tvindex + 1 : tvindex;
+      }
+      else {
+        parsePacket(currentPacket, mPacketFirstTs, traceVector.mArray[tvindex++]);
+      }
+      traceVector.mLength = tvindex;
+    } // For i < count
+    mclockTrainingdone = true;
+}
+
+  // Convert decimal to binary string
+  // NOTE: length of string is always sizeof(uint32_t) * 8
+  std::string TraceS2MM::dec2bin(uint32_t n) {
+    char result[(sizeof(uint32_t) * 8) + 1];
```

Comment:
Would it be more efficient to allocate the string up front using std::string(count, 0) then get the char* pointer using data()?  That way RVO (return value optimization) can kick in.

Reply:
updated this function